File: solutions/csharp/protein-translation/1/ProteinTranslation.cs
using System;
using System.Collections.Generic;

public static class ProteinTranslation
{
    private static string GetProtein(string codon)
    {
        switch (codon)
        {
            case "AUG":
                return "Methionine";
            case "UUU":
            case "UUC":
                return "Phenylalanine";
            case "UUA":
            case "UUG":
                return "Leucine";
            case "UCU":
            case "UCC":
            case "UCA":
            case "UCG":
                return "Serine";
            case "UAU":
            case "UAC":
                return "Tyrosine";
            case "UGU":
            case "UGC":
                return "Cysteine";
            case "UGG":
                return "Tryptophan";
            case "UAA":
            case "UAG":
            case "UGA":
            default:
                return string.Empty;
        }
    }
    public static string[] Proteins(string strand)
    {
        List<string> proteins = new();
        int index = 0;
        while (index < strand.Length)
        {
            string protein = GetProtein(strand.Substring(index, 3));
            if (string.IsNullOrEmpty(protein))
            {
                break;
            }
            proteins.Add(protein);
            index += 3;
        }

        return proteins.ToArray();
    }
}

File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"
#include <string>
using std::string;
#include <vector>
using std::vector;

namespace protein_translation {
vector<string> proteins(string codons) {
  vector<string> result;
  for (int i = 0; i < static_cast<int>(codons.length()); i += 3) {
    if (i + 2 < static_cast<int>(codons.length())) {
      string codon = codons.substr(i, 3);
      string protein = get_protein(codon);
      if (protein == "STOP") {
        break;
      }
      if (protein != "") {
        result.push_back(protein);
      }
    }
  }
  return result;
}
string get_protein(string &codon) {
  if (codon == "AUG") {
    return "Methionine";
  }
  if (codon == "UUU" || codon == "UUC") {
    return "Phenylalanine";
  }
  if (codon == "UUA" || codon == "UUG") {
    return "Leucine";
  }
  if (codon == "UCU" || codon == "UCC" || codon == "UCA" || codon == "UCG") {
    return "Serine";
  }
  if (codon == "UAU" || codon == "UAC") {
    return "Tyrosine";
  }
  if (codon == "UGU" || codon == "UGC") {
    return "Cysteine";
  }
  if (codon == "UGG") {
    return "Tryptophan";
  }
  if (codon == "UAA" || codon == "UAG" || codon == "UGA") {
    return "STOP";
  }
  return "";
}
} // namespace protein_translation
